File: srcs/juloo.keyboard2/NeuralSwipeTypingEngine.java
package juloo.keyboard2;

import android.content.Context;
import android.graphics.PointF;
import android.util.Log;
import java.util.Map;

/**
 * Neural swipe typing engine using ONNX transformer models
 * Replaces SwipeTypingEngine with neural prediction pipeline
 * Maintains same interface for seamless integration
 */
public class NeuralSwipeTypingEngine
{
  public interface DebugLogger
  {
    void log(String message);
  }
  private static final String TAG = "NeuralSwipeTypingEngine";
  
  private Context _context;
  private Config _config;
  private OnnxSwipePredictor _neuralPredictor;
  
  // State tracking
  private boolean _initialized = false;
  
  // Debug logging callback
  private DebugLogger _debugLogger;
  
  public NeuralSwipeTypingEngine(Context context, Config config)
  {
    _context = context;
    _config = config;
    
    // OPTIMIZATION: Use singleton predictor with session persistence
    _neuralPredictor = OnnxSwipePredictor.getInstance(context);
    
    Log.d(TAG, "NeuralSwipeTypingEngine created - using persistent singleton predictor");
  }
  
  /**
   * Initialize the engine and load models
   */
  public boolean initialize()
  {
    if (_initialized)
    {
      return true;
    }
    
    try
    {
      Log.d(TAG, "Initializing pure neural swipe engine...");
      
      // Propagate debug logger to predictor
      if (_debugLogger != null)
      {
        _neuralPredictor.setDebugLogger(_debugLogger);
      }
      
      // Initialize neural predictor - MUST succeed or throw error
      // Note: Singleton may already be initialized, which is optimal for performance
      boolean neuralReady = _neuralPredictor.initialize();
      
      if (!neuralReady)
      {
        throw new RuntimeException("Failed to initialize ONNX neural models");
      }
      
      _initialized = true;
      
      Log.d(TAG, "Neural engine initialized successfully - pure neural mode");
      
      return true;
    }
    catch (Exception e)
    {
      Log.e(TAG, "Failed to initialize neural engine", e);
      // Pure neural - no fallback variables needed
      _initialized = true;
      return false;
    }
  }
  
  /**
   * Main prediction method - maintains compatibility with legacy interface
   */
  public PredictionResult predict(SwipeInput input)
  {
    // Add stack trace to see who's calling this
    Log.d(TAG, "🔥🔥🔥 NEURAL PREDICTION CALLED FROM:");
    for (StackTraceElement element : Thread.currentThread().getStackTrace()) {
      if (element.getClassName().contains("juloo.keyboard2")) {
        Log.d(TAG, "🔥   " + element.getClassName() + "." + element.getMethodName() + ":" + element.getLineNumber());
      }
    }
    
    if (!_initialized)
    {
      initialize();
    }
    
    Log.d(TAG, "=== PURE NEURAL PREDICTION START ===");
    Log.d(TAG, String.format(
      "Input: keySeq=%s, pathLen=%.1f, duration=%.2fs", 
      input.keySequence, input.pathLength, input.duration));
    
    Log.d(TAG, "Using PURE NEURAL prediction - no classification needed");
    
    try
    {
      PredictionResult result = _neuralPredictor.predict(input);
      
      if (result != null)
      {
        Log.d(TAG, String.format("Neural prediction successful: %d candidates", 
          result.words.size()));
        return result;
      }
      else
      {
        throw new RuntimeException("Neural prediction returned null result");
      }
    }
    catch (Exception e)
    {
      Log.e(TAG, "Neural prediction failed", e);
      throw new RuntimeException("Neural prediction failed: " + e.getMessage());
    }
  }
  
  /**
   * Set keyboard dimensions for coordinate mapping
   */
  public void setKeyboardDimensions(float width, float height)
  {
    if (_neuralPredictor != null)
    {
      _neuralPredictor.setKeyboardDimensions(width, height);
      Log.d(TAG, String.format("Set keyboard dimensions: %.0fx%.0f", width, height));
    }
  }
  
  /**
   * Set real key positions for accurate coordinate mapping
   */
  public void setRealKeyPositions(Map<Character, PointF> realPositions)
  {
    if (_neuralPredictor != null)
    {
      _neuralPredictor.setRealKeyPositions(realPositions);
      Log.d(TAG, String.format("Set key positions: %d keys", 
        realPositions != null ? realPositions.size() : 0));
    }
  }
  
  /**
   * Update configuration
   */
  public void setConfig(Config config)
  {
    _config = config;
    
    // Update neural predictor configuration
    if (_neuralPredictor != null)
    {
      _neuralPredictor.setConfig(config);
    }
    
    Log.d(TAG, "Neural config updated");
  }
  
  /**
   * Check if neural prediction is available
   */
  public boolean isNeuralAvailable()
  {
    return _neuralPredictor != null && _neuralPredictor.isAvailable();
  }
  
  /**
   * Get current prediction mode
   */
  public String getCurrentMode()
  {
    return isNeuralAvailable() ? "neural" : "error";
  }
  
  /**
   * Set debug logger for detailed logging
   */
  public void setDebugLogger(DebugLogger logger)
  {
    _debugLogger = logger;
    if (_neuralPredictor != null)
    {
      _neuralPredictor.setDebugLogger(logger);
    }
  }
  
  private void logDebug(String message)
  {
    if (_debugLogger != null)
    {
      _debugLogger.log(message);
    }
  }
  
  /**
   * Clean up resources
   */
  public void cleanup()
  {
    if (_neuralPredictor != null)
    {
      _neuralPredictor.cleanup();
    }
    
    Log.d(TAG, "Neural swipe engine cleaned up");
  }
}

File: srcs/juloo.keyboard2/NeuralSwipeTypingEngine.kt
package juloo.keyboard2

import android.content.Context
import android.graphics.PointF
import android.util.Log
import juloo.keyboard2.onnx.PredictionPostProcessor
import juloo.keyboard2.onnx.SwipePredictorOrchestrator

/**
 * Neural swipe typing engine using ONNX transformer models
 * Replaces SwipeTypingEngine with neural prediction pipeline
 * Maintains same interface for seamless integration
 */
class NeuralSwipeTypingEngine(
    private val context: Context,
    private var config: Config
) {
    fun interface DebugLogger {
        fun log(message: String)
    }

    companion object {
        private const val TAG = "NeuralSwipeTypingEngine"
    }

    private val neuralPredictor: SwipePredictorOrchestrator

    // State tracking
    private var initialized = false

    // Debug logging callback
    private var debugLogger: DebugLogger? = null

    init {
        // OPTIMIZATION: Use singleton predictor with session persistence
        neuralPredictor = SwipePredictorOrchestrator.getInstance(context)

        // OPTIMIZATION: Start async model loading immediately for faster startup
        // Models will load in background while keyboard UI appears
        neuralPredictor.initializeAsync()

        Log.d(TAG, "NeuralSwipeTypingEngine created - using persistent singleton predictor")
    }

    /**
     * Initialize the engine and load models
     */
    fun initialize(): Boolean {
        if (initialized) {
            return true
        }

        return try {
            Log.d(TAG, "Initializing pure neural swipe engine...")

            // Propagate debug logger to predictor
            debugLogger?.let {
                neuralPredictor.setDebugLogger(it)
            }

            // Initialize neural predictor - MUST succeed or throw error
            // Note: Singleton may already be initialized, which is optimal for performance
            val neuralReady = neuralPredictor.initialize()

            if (!neuralReady) {
                throw RuntimeException("Failed to initialize ONNX neural models")
            }

            initialized = true

            Log.d(TAG, "Neural engine initialized successfully - pure neural mode")

            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize neural engine", e)
            // Pure neural - no fallback variables needed
            initialized = true
            false
        }
    }

    /**
     * Main prediction method - maintains compatibility with legacy interface
     */
    fun predict(input: SwipeInput): PredictionResult {
        // Add stack trace to see who's calling this
        Log.d(TAG, "🔥🔥🔥 NEURAL PREDICTION CALLED FROM:")
        for (element in Thread.currentThread().stackTrace) {
            if (element.className.contains("juloo.keyboard2")) {
                Log.d(TAG, "🔥   ${element.className}.${element.methodName}:${element.lineNumber}")
            }
        }

        if (!initialized) {
            initialize()
        }

        Log.d(TAG, "=== PURE NEURAL PREDICTION START ===")
        Log.d(TAG, "Input: keySeq=${input.keySequence}, pathLen=${"%.1f".format(input.pathLength)}, " +
            "duration=${"%.2f".format(input.duration)}s")

        Log.d(TAG, "Using PURE NEURAL prediction - no classification needed")

        return try {
            val result = neuralPredictor.predict(input)

            if (result != null) {
                val words = result.words
                val scores = result.scores

                Log.d(TAG, "Neural prediction successful: ${words.size} candidates")
                PredictionResult(words, scores)
            } else {
                throw RuntimeException("Neural prediction returned null result")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Neural prediction failed", e)
            throw RuntimeException("Neural prediction failed: ${e.message}")
        }
    }

    /**
     * Set keyboard dimensions for coordinate mapping
     */
    fun setKeyboardDimensions(width: Float, height: Float) {
        neuralPredictor.setKeyboardDimensions(width, height)
        Log.d(TAG, "Set keyboard dimensions: ${"%.0f".format(width)}x${"%.0f".format(height)}")
    }

    /**
     * Set QWERTY area bounds for proper coordinate normalization.
     * This is critical for correct key detection - the model expects coordinates
     * normalized over just the QWERTY key area, not the full keyboard view.
     *
     * @param qwertyTop Y offset in pixels where QWERTY keys start
     * @param qwertyHeight Height in pixels of the QWERTY key area
     */
    fun setQwertyAreaBounds(qwertyTop: Float, qwertyHeight: Float) {
        neuralPredictor.setQwertyAreaBounds(qwertyTop, qwertyHeight)
        Log.d(TAG, "Set QWERTY area bounds: top=${"%.0f".format(qwertyTop)}, " +
            "height=${"%.0f".format(qwertyHeight)}")
    }

    /**
     * Set touch Y-offset compensation for fat finger effect.
     * Users typically touch above key centers; this offset compensates.
     *
     * @param offset Pixels to add to Y coordinate (positive = shift down)
     */
    fun setTouchYOffset(offset: Float) {
        neuralPredictor.setTouchYOffset(offset)
        Log.d(TAG, "Set touch Y-offset: ${"%.0f".format(offset)} pixels")
    }

    /**
     * Set real key positions for accurate coordinate mapping
     */
    fun setRealKeyPositions(realPositions: Map<Char, PointF>?) {
        neuralPredictor.setRealKeyPositions(realPositions)
        Log.d(TAG, "Set key positions: ${realPositions?.size ?: 0} keys")
    }

    /**
     * Update configuration
     */
    fun setConfig(config: Config) {
        this.config = config

        // Update neural predictor configuration
        neuralPredictor.setConfig(config)

        Log.d(TAG, "Neural config updated")
    }

    /**
     * Check if neural prediction is available
     */
    fun isNeuralAvailable(): Boolean {
        return neuralPredictor.isAvailable()
    }

    /**
     * Get current prediction mode
     */
    fun getCurrentMode(): String {
        return if (isNeuralAvailable()) "neural" else "error"
    }

    /**
     * Set debug logger for detailed logging
     */
    fun setDebugLogger(logger: DebugLogger?) {
        debugLogger = logger
        debugLogger?.let {
            neuralPredictor.setDebugLogger(it)
        }
    }

    private fun logDebug(message: String) {
        debugLogger?.log(message)
    }

    /**
     * Clean up resources
     */
    fun cleanup() {
        neuralPredictor.cleanup()

        Log.d(TAG, "Neural swipe engine cleaned up")
    }
}
